File: .claude/worktrees/vibrant-pike-750673/services/enrichment/workers/spotify.js
'use strict';
// ── Spotify enrichment worker ─────────────────────────────────────────────────
// API: https://api.spotify.com/v1  — Client Credentials flow (geen user-auth)
// Rate limit: 1 req/sec conservatief; respecteer Retry-After headers
// Vereist: SPOTIFY_CLIENT_ID + SPOTIFY_CLIENT_SECRET in process.env

const RATE_INTERVAL = 1100; // ms tussen calls (iets boven 1/sec)
const FUZZY_THRESHOLD = 0.80;
const DAILY_CAP = 3000;     // max Spotify-calls per dag

// ── Token-beheer (gedeeld met bestaande spotify.js service) ──────────────────
let _token    = null;
let _tokenExp = 0;
let _dailyCount = 0;
let _dailyDate  = '';

async function _getToken() {
  if (_token && Date.now() < _tokenExp - 5 * 60_000) return _token;

  const CLIENT_ID     = process.env.SPOTIFY_CLIENT_ID;
  const CLIENT_SECRET = process.env.SPOTIFY_CLIENT_SECRET;

  if (!CLIENT_ID || !CLIENT_SECRET) throw new Error('Spotify credentials niet geconfigureerd');

  const creds = Buffer.from(`${CLIENT_ID}:${CLIENT_SECRET}`).toString('base64');
  const res   = await fetch('https://accounts.spotify.com/api/token', {
    method:  'POST',
    headers: {
      Authorization:  `Basic ${creds}`,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body:   'grant_type=client_credentials',
    signal: AbortSignal.timeout(10_000),
  });

  if (!res.ok) throw new Error(`Spotify token fout ${res.status}`);
  const data  = await res.json();
  _token      = data.access_token;
  _tokenExp   = Date.now() + data.expires_in * 1000;
  return _token;
}

// ── Fuzzy string matching (Dice-coëfficiënt, vergelijkbaar met SequenceMatcher) ──
function _fuzzyScore(a, b) {
  a = (a || '').toLowerCase().trim();
  b = (b || '').toLowerCase().trim();
  if (a === b) return 1.0;
  if (a.length < 2 || b.length < 2) return 0.0;

  const bigrams = (s) => {
    const m = new Map();
    for (let i = 0; i < s.length - 1; i++) {
      const bg = s.slice(i, i + 2);
      m.set(bg, (m.get(bg) || 0) + 1);
    }
    return m;
  };

  const aMap = bigrams(a);
  const bMap = bigrams(b);
  let intersection = 0;

  for (const [bg, cnt] of aMap) {
    const bCnt = bMap.get(bg) || 0;
    intersection += Math.min(cnt, bCnt);
  }

  const total = a.length + b.length - 2;
  return total > 0 ? (2 * intersection) / total : 0;
}

class SpotifyWorker {
  constructor(db, log) {
    this.db        = db;
    this.log       = log.child ? log.child({ worker: 'spotify' }) : log;
    this._lastCall = 0;
  }

  // ── Rate limiting ──────────────────────────────────────────────────────────

  async _rateLimit(retryAfter = 0) {
    const wait = Math.max(retryAfter * 1000, RATE_INTERVAL - (Date.now() - this._lastCall));
    if (wait > 0) await new Promise(r => setTimeout(r, wait));
    this._lastCall = Date.now();
  }

  _checkDailyBudget() {
    const today = new Date().toISOString().slice(0, 10);
    if (_dailyDate !== today) { _dailyDate = today; _dailyCount = 0; }
    if (_dailyCount >= DAILY_CAP) throw new Error(`Spotify daily budget van ${DAILY_CAP} bereikt`);
    _dailyCount++;
  }

  async _get(path, params = {}) {
    this._checkDailyBudget();
    await this._rateLimit();

    const token = await _getToken();
    const url   = new URL(`https://api.spotify.com${path}`);
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
    }

    const res = await fetch(url.toString(), {
      headers: { Authorization: `Bearer ${token}`, Accept: 'application/json' },
      signal:  AbortSignal.timeout(12_000),
    });

    // Respecteer Retry-After bij 429
    if (res.status === 429) {
      const retryAfter = parseInt(res.headers.get('Retry-After') || '5', 10);
      this.log.warn({ retryAfter }, 'Spotify rate limit — wacht');
      await this._rateLimit(retryAfter);
      return this._get(path, params); // één retry
    }

    if (!res.ok) throw new Error(`Spotify HTTP ${res.status} bij ${path}`);
    return res.json();
  }

  // ── Hoofdverwerker ────────────────────────────────────────────────────────

  async process(entity) {
    try {
      const type = entity.entity_type;
      let data;

      if (type === 'artist') {
        data = await this._processArtist(entity.entity_name);
      } else if (type === 'album') {
        data = await this._processAlbum(entity.entity_name);
      } else if (type === 'track') {
        data = await this._processTrack(entity.entity_name);
      } else {
        return { ok: false, error: `Onbekend entity_type: ${type}` };
      }

      if (!data) return { ok: false, error: 'Geen resultaat gevonden' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'Spotify worker mislukt');
      return { ok: false, error: err.message };
    }
  }

  // ── Artiest ───────────────────────────────────────────────────────────────

  async _processArtist(name) {
    const results = await this._get('/v1/search', { q: name, type: 'artist', limit: 5 });
    const items   = results.artists?.items || [];
    if (!items.length) return null;

    const best = this._fuzzyBest(items, name, a => a.name);
    if (!best) return null;

    return {
      spotify_id:   best.id,
      genres:       best.genres || [],
      popularity:   best.popularity ?? null,
      artwork_url:  best.images?.[0]?.url || null,
      spotify_url:  best.external_urls?.spotify || null,
      followers:    best.followers?.total ?? null,
      source:       'spotify',
      fetchedAt:    Date.now(),
    };
  }

  // ── Album ─────────────────────────────────────────────────────────────────

  async _processAlbum(name) {
    const results = await this._get('/v1/search', { q: name, type: 'album', limit: 5 });
    const items   = results.albums?.items || [];
    if (!items.length) return null;

    const best = this._fuzzyBest(items, name, a => a.name);
    if (!best) return null;

    return {
      spotify_id:   best.id,
      artwork_url:  best.images?.[0]?.url || null,
      genres:       best.genres || [],
      release_date: best.release_date || null,
      total_tracks: best.total_tracks ?? null,
      spotify_url:  best.external_urls?.spotify || null,
      source:       'spotify',
      fetchedAt:    Date.now(),
    };
  }

  // ── Track ─────────────────────────────────────────────────────────────────

  async _processTrack(name) {
    const results = await this._get('/v1/search', { q: name, type: 'track', limit: 5 });
    const items   = results.tracks?.items || [];
    if (!items.length) return null;

    const best = this._fuzzyBest(items, name, t => t.name);
    if (!best) return null;

    return {
      spotify_id:   best.id,
      preview_url:  best.preview_url || null,
      popularity:   best.popularity ?? null,
      duration_ms:  best.duration_ms ?? null,
      explicit:     best.explicit ?? null,
      artwork_url:  best.album?.images?.[0]?.url || null,
      spotify_url:  best.external_urls?.spotify || null,
      source:       'spotify',
      fetchedAt:    Date.now(),
    };
  }

  // ── Hulpfuncties ──────────────────────────────────────────────────────────

  /**
   * Kies het beste item uit een lijst via fuzzy matching.
   * @param {Array}    items
   * @param {string}   query
   * @param {Function} nameGetter  item → string
   * @returns het beste item, of null als onder threshold
   */
  _fuzzyBest(items, query, nameGetter) {
    // Eerst exacte match (case-insensitive)
    const norm  = s => (s || '').toLowerCase().trim();
    const exact = items.find(i => norm(nameGetter(i)) === norm(query));
    if (exact) return exact;

    // Dan fuzzy
    let best = null;
    let bestScore = 0;
    for (const item of items) {
      const score = _fuzzyScore(nameGetter(item), query);
      if (score > bestScore) { bestScore = score; best = item; }
    }

    if (bestScore >= FUZZY_THRESHOLD) return best;
    return null;
  }

  /** Valideer dat een Spotify ID correct is (alfanumeriek, 22 tekens). */
  _isValidSpotifyId(id) {
    return typeof id === 'string' && /^[a-zA-Z0-9]{22}$/.test(id);
  }
}

module.exports = { SpotifyWorker };

File: core/workers/spotify.py
"""
spotify.py — Spotify enrichment worker.

Gebruikt spotipy met Client Credentials flow (geen user-auth nodig).
Vereist SPOTIFY_CLIENT_ID + SPOTIFY_CLIENT_SECRET.
Rate limit: conservatief 1 call/sec + respecteer Retry-After bij 429.
"""
from __future__ import annotations

import logging
import os
import time
from typing import Any

import spotipy
from spotipy.oauth2 import SpotifyClientCredentials

from core.workers.base import BaseWorker

FUZZY_THRESHOLD = 0.80
DAILY_CAP       = 3000
RATE_INTERVAL   = 1.1   # seconden


class SpotifyWorker(BaseWorker):
    source = "spotify"

    def __init__(self, database, logger: logging.Logger):
        super().__init__(database, logger)
        self._last_call   = 0.0
        self._daily_count = 0
        self._daily_date  = ""
        self._sp: spotipy.Spotify | None = None
        self._init_client()

    def _init_client(self) -> None:
        client_id     = os.environ.get("SPOTIFY_CLIENT_ID", "")
        client_secret = os.environ.get("SPOTIFY_CLIENT_SECRET", "")
        if client_id and client_secret:
            ccm = SpotifyClientCredentials(
                client_id=client_id,
                client_secret=client_secret,
            )
            self._sp = spotipy.Spotify(
                auth_manager=ccm,
                requests_timeout=12,
                retries=1,
                backoff_factor=0.5,
            )

    @property
    def is_configured(self) -> bool:
        return self._sp is not None

    def _rate_limit(self) -> None:
        wait = max(0.0, RATE_INTERVAL - (time.time() - self._last_call))
        if wait > 0:
            time.sleep(wait)
        self._last_call = time.time()

    def _check_daily_budget(self) -> None:
        today = time.strftime("%Y-%m-%d")
        if self._daily_date != today:
            self._daily_date  = today
            self._daily_count = 0
        if self._daily_count >= DAILY_CAP:
            raise RuntimeError(f"Spotify daily budget van {DAILY_CAP} bereikt")
        self._daily_count += 1

    def _search(self, query: str, search_type: str, limit: int = 5) -> dict:
        self._check_daily_budget()
        self._rate_limit()
        return self._sp.search(q=query, type=search_type, limit=limit) or {}

    def process(self, item: dict) -> dict[str, Any]:
        if not self.is_configured:
            return {"ok": False, "error": "Spotify credentials niet geconfigureerd"}
        try:
            entity_type = item["entity_type"]
            name        = item["entity_name"]

            if entity_type == "artist":
                data = self._process_artist(name)
            elif entity_type == "album":
                data = self._process_album(name)
            elif entity_type == "track":
                data = self._process_track(name)
            else:
                return {"ok": False, "error": f"Onbekend entity_type: {entity_type}"}

            if not data:
                return {"ok": False, "error": "Geen resultaat gevonden"}
            return {"ok": True, "data": data}

        except Exception as exc:
            self.log.warning(f"Spotify worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    def _process_artist(self, name: str) -> dict | None:
        result = self._search(name, "artist")
        items  = (result.get("artists") or {}).get("items", [])
        if not items:
            return None

        best = self.fuzzy_best(items, name, lambda a: a.get("name", ""), FUZZY_THRESHOLD)
        if not best:
            return None

        images = best.get("images") or []
        return {
            "spotify_id":  best["id"],
            "genres":      best.get("genres", []),
            "popularity":  best.get("popularity"),
            "artwork_url": images[0]["url"] if images else None,
            "spotify_url": (best.get("external_urls") or {}).get("spotify"),
            "followers":   (best.get("followers") or {}).get("total"),
            "source":      "spotify",
            "fetchedAt":   int(time.time() * 1000),
        }

    def _process_album(self, name: str) -> dict | None:
        result = self._search(name, "album")
        items  = (result.get("albums") or {}).get("items", [])
        if not items:
            return None

        best = self.fuzzy_best(items, name, lambda a: a.get("name", ""), FUZZY_THRESHOLD)
        if not best:
            return None

        images = best.get("images") or []
        return {
            "spotify_id":   best["id"],
            "artwork_url":  images[0]["url"] if images else None,
            "genres":       best.get("genres", []),
            "release_date": best.get("release_date"),
            "total_tracks": best.get("total_tracks"),
            "spotify_url":  (best.get("external_urls") or {}).get("spotify"),
            "source":       "spotify",
            "fetchedAt":    int(time.time() * 1000),
        }

    def _process_track(self, name: str) -> dict | None:
        result = self._search(name, "track")
        items  = (result.get("tracks") or {}).get("items", [])
        if not items:
            return None

        best = self.fuzzy_best(items, name, lambda t: t.get("name", ""), FUZZY_THRESHOLD)
        if not best:
            return None

        album_images = (best.get("album") or {}).get("images") or []
        return {
            "spotify_id":  best["id"],
            "preview_url": best.get("preview_url"),
            "popularity":  best.get("popularity"),
            "duration_ms": best.get("duration_ms"),
            "explicit":    best.get("explicit"),
            "artwork_url": album_images[0]["url"] if album_images else None,
            "spotify_url": (best.get("external_urls") or {}).get("spotify"),
            "source":      "spotify",
            "fetchedAt":   int(time.time() * 1000),
        }
